### DISCUSSION TITLE
Support for Complex QuantityValue

### DISCUSSION THREAD

**Discussion by mwagnerEE** [angularsen/UnitsNet#1048] [Ideas]
Have you ever considered implementing the `System.Numerics.Complex` data type? Instead of some quantities using `decimal` and some using `double`, could they not all use `QuantityValue`? And in that event, what would prevent the extension of this struct to have a third field to store `Complex` as like this:

```cs

private QuantityValue(Complex val)
{
    _valueComplex = val;
    _valueDecimal = null;
    _value = null;
}

#region To QuantityValue
/// <summary>Implicit cast from <see cref="Complex"/> to <see cref="QuantityValue"/>.</summary>
public static implicit operator QuantityValue(Complex val) => new QuantityValue(val); // Complex
#endregion

#region To Complex

/// <summary>Explicit cast from <see cref="QuantityValue"/> to <see cref="Complex"/>.</summary>
public static explicit operator Complex(QuantityValue number)
{
    return number._valueComplex ?? new Complex(number._value ?? (double)number._valueDecimal.GetValueOrDefault(), 0);
}

#endregion

/// <summary>Returns the string representation of the numeric value.</summary>
public override string ToString()
{
    if (_value.HasValue)
    {
        return _value.ToString();
    }
    else if (_valueDecimal.HasValue)
    {
        return _valueDecimal.ToString();
    }
    else if (_valueComplex.HasValue)
    {
        return _valueComplex.ToString();
    }
    else
    {
        return "0";
        //throw new System.Exception();
    }
}

```

**Comment by angularsen** [comment]
Hi, Complex type has not been discussed before. Do you have any examples of using complex numbers to represent a quantity value? Would you use it if it was implemented?

As for using QuantityValue as the internal value, maybe. I haven't really thought about it, but maybe there is some merit to it. I'm trying to come up with counter-arguments not to do so, and I think the best I can think of right now is that serialization will be more complicated and it probably adds some extra bytes for memory sensitive applications.

**Reply by mwagnerEE** [reply]
I'm an EE and I'm trying to make a power flow analyzer. Complex values are used all the time in Electrical. In messing around with allowing arbitrary arithmetic of quantities, I did implement Complex values. Although I replaced decimal and double for ease of prototyping. I ended up just copying DotNet's `Complex` and extending it. It probably would take up more space but is otherwise hidden in that a `double` is implicitly convertible to `Complex` so if you didn't use the imaginary parts you could continue to use `double`s

**Reply by angularsen** [reply]
Related: #1084 

I don't think Complex fits into that solution though, we would probably have to extend QuantityValue somehow to make it work. That would probably also ramp up the byte size again and I would like to avoid that.

At this point, it feels like Complex values may be out of scope for UnitsNet. If you see a clever way forward I'm open to it, but it may simply be better solved outside.
